### PR TITLE
[FEATURE] aligned_allocator

### PR DESCRIPTION
**Pull request by marehr**
This PR is needed for storing SIMD vectors in a container (like `std::vector`) with a correct memory alignment.

For example SIMD vectors for `sse4` are `128` bit = `8byte` long and need an memory alignment of `128 bit` = `8 byte`. 

Example
```c++
size_t memory_alignment(void * value, size_t alignment)
{
   return (((size_t)value) & (alignment-1));
}

int main()
{
    // 128-byte memory aligned and 16bit = 2byte address width for each element
    std::vector<int16_t, aligned_allocator<int16_t, 128u>> vec128{1,2,3,4,5};

    // vector has no alignment and 16bit = 2byte address width for each element
    std::vector<int16_t> vec_unaligned{1,2,3,4,5};

    // 256-byte memory aligned and 32bit = 4byte address width for each element
    std::vector<int32_t, aligned_allocator<int32_t, 256u>> vec256{1,2,3,4,5};

    for (auto && x: vec128)
        std::cout << "Item: " << x << " (" << &x << ", 128-byte aligned offset: " << memory_alignment(&x, 128u) << ")" << std::endl;

    for (auto && x: vec_unaligned)
        std::cout << "Item: " << x << " (" << &x << ", unaligned start: " << memory_alignment(&x, 128u) << ")" << std::endl;

    for (auto && x: vec256)
        std::cout << "Item: " << x << " (" << &x << ", 256-byte aligned offset: " << memory_alignment(&x, 256u) << ")" << std::endl;
}
```

```terminal
 * Item: 1 (0x55d5d0722f00, 128-byte aligned offset: 0)
 * Item: 2 (0x55d5d0722f02, 128-byte aligned offset: 2)
 * Item: 3 (0x55d5d0722f04, 128-byte aligned offset: 4)
 * Item: 4 (0x55d5d0722f06, 128-byte aligned offset: 6)
 * Item: 5 (0x55d5d0722f08, 128-byte aligned offset: 8)
 * Item: 1 (0x55d5d0722f40, unaligned start: 64)
 * Item: 2 (0x55d5d0722f42, unaligned start: 66)
 * Item: 3 (0x55d5d0722f44, unaligned start: 68)
 * Item: 4 (0x55d5d0722f46, unaligned start: 70)
 * Item: 5 (0x55d5d0722f48, unaligned start: 72)
 * Item: 1 (0x55d5d0723000, 256-byte aligned offset: 0)
 * Item: 2 (0x55d5d0723004, 256-byte aligned offset: 4)
 * Item: 3 (0x55d5d0723008, 256-byte aligned offset: 8)
 * Item: 4 (0x55d5d072300c, 256-byte aligned offset: 12)
 * Item: 5 (0x55d5d0723010, 256-byte aligned offset: 16)
```

### CODE REVIEW

File: include/seqan3/memory/aligned_allocator.hpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+/*!\file
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ * \brief Contains seqan3::aligned_allocator.
+ */
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <limits>
```
Comment (h-2) [inline include/seqan3/memory/aligned_allocator.hpp]: alphabetical order and add seqan3/platform.hpp

**Comment by marehr** [comment]
Wasn't sure where to put it. Will move it :heart:

File: test/snippet/range/container/aligned_allocator.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#include <iostream>
+#include <vector>
+
+#include <seqan3/range/container/aligned_allocator.hpp>
+
+size_t memory_alignment(void * value, size_t alignment)
+{
+   return (((size_t)value) & (alignment-1));
```
Comment (h-2) [inline test/snippet/range/container/aligned_allocator.cpp]: don't use c-style casts :nail_care:

Comment (h-2) [reply test/snippet/range/container/aligned_allocator.cpp]: and space before and after `-`

File: test/snippet/range/container/aligned_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    std::vector<int32_t, aligned_allocator<int32_t, 256u>> vec256{1,2,3,4,5};
+
+    for (auto && x: vec128)
+        std::cout << "Item: " << x << " (" << &x << ", 128-byte aligned offset: " << memory_alignment(&x, 128u) << ")" << std::endl;
+
+    for (auto && x: vec_unaligned)
+        std::cout << "Item: " << x << " (" << &x << ", unaligned start: " << memory_alignment(&x, 128u) << ")" << std::endl;
+
+    for (auto && x: vec256)
+        std::cout << "Item: " << x << " (" << &x << ", 256-byte aligned offset: " << memory_alignment(&x, 256u) << ")" << std::endl;
```
Comment (h-2) [inline test/snippet/range/container/aligned_allocator.cpp]: as a way of educating our users, also prefer `\n` to `std::endl`.

**Comment by h-2** [comment]
!rebuild

**Comment by h-2** [comment]
what happened to jenkins?

**Pull request merged**